---
layout: page
title: Variants/genes annotation and prioritization
summary: "Variants/genes annotation and prioritization"
sidebar_link: true
---


Once annotated, variants need to be filtered and prioritized

  - WES: Tens of thousands
  - WGS: Millions

**No universal filters, they depend on the experimental features**

---
### Variant impact

First of all you usually want to filter variants by consequence on the encoded protein, keeping those which have an higher impact on protein:
 
- Missense
- Nonsense
- Splice sites
- Frameshift indels
- Inframe indels

---
### Variant frequency
 
- Common variants are unlikely associated with a clinical condition
- A rare variant will probably have a higher functional effect on the protein
- Frequency cut-off have to be customized on each different case
- Typical cut-offs: 1% - 0.1%
- Allele frequencies may differ a lot between different populations

---
### Variant effect prediction Tools

- Tools that predict consequences of amino acid substitutions on protein function
- They give a score and/or a prediction in terms of "Tolerated", "Deleterious" (SIFT) or "Probably Damaging", "Possibly Damaging", "Benign" (Polyphen2)
  
  - fitCons
  - GERP++
  - SIFT
  - PolyPhen2
  - CADD
  - DANN
  - Condel
  - fathmm
  - MutationTaster
  - MutationAssessor
  - REVEL

---
### ACMG/AMP 2015 guidelines

The *American College of Medical Genetics* and the *Association for Molecular Pathology* published guidelines for the interpretation of sequence variants in May of 2015 [(Richards S. et al, 2015)](https://www.nature.com/articles/gim201530). This report describes updated standards and guidelines for classifying sequence variants by using criteria informed by expert opinion and experience

- 28 evaluation criteria for the clinical interpretation of variants
  - Criteria falls into 3 sets:
    - pathogenic/likely pathogenic (P/LP)
    - benign/likely benign (B/LB)
    - variant of unknown significance (VUS)

- [**Intervar**](http://wintervar.wglab.org): software for automatically interpretation of the 28 criteria
  - Two major steps:
    - automatically interpretation by 28 criteria
    - manual adjustment to re-interpret the clinical significance

---
### Prioritization

Phenotype-based prioritization tools are methods working by comparing the phenotypes of a patient with gene-phenotype known associations.

- [Phenolyzer](http://phenolyzer.wglab.org/): Phenotype Based Gene Analyzer, a tool to prioritize genes based on user-specific disease/phenotype terms

---
## 1. Annotation and filtering training

Using knowledge gained on [Genomic databases and variant annotation]({{site.url}}{{site.baseurl}}/lectures/04.db.html) section, try to ***annotate***, ***filter*** and ***prioritize*** an example exome variant data, using two disease terms

- Download file Sample1.all_exons.hg19.vcf from the *dataset*, as learned in the [Home]({{site.url}}{{site.baseurl}}/lectures/01.home.html) section

- Go to [wANNOVAR](http://wannovar.wglab.org/)

- Use the vcf file as **input file**

- Use *hearing loss* and *deafness autosomal recessive*, as disease terms to prioritize results

- Choose *rare recessive Mendelian disease* as Disease Model in the ***Parameter Settings*** section

- provide an istitutional email address and submit the Job


... *wait for results* ...


- In the results page you can navigate and download results

- Click ***Show*** in the *Network Visualization* section to see Phenolyzer prioritization results
